test/unit_test/passes/onnx/test_graph_surgeries.py

```diff
@@ -1731,3 +1731,91 @@ def test_quickgelu_to_sigmoid(tmp_path):
     assert OpType.Mul in op_types
     assert OpType.Sigmoid in op_types
     assert op_types.count(OpType.Mul) == 2
+
+
+def test_lower_rotary_embedding_numerical(tmp_path):
+    """Test LowerRotaryEmbedding with numerical validation."""
+    # setup
+    batch_size, seq_len, hidden_size = 2, 8, 64
+    max_seq_len = 128
+    rotary_dim = hidden_size // 2
+
+    # Generate position embeddings for cos/sin caches
+    position = np.arange(max_seq_len)[:, np.newaxis]
+    div_term = np.exp(np.arange(0, rotary_dim, 2) * -(np.log(10000.0) / rotary_dim))
+
+    cos_cache = np.zeros((max_seq_len, rotary_dim), dtype=np.float32)
+    sin_cache = np.zeros((max_seq_len, rotary_dim), dtype=np.float32)
+    cos_cache[:, 0::2] = np.cos(position * div_term)
+    cos_cache[:, 1::2] = np.cos(position * div_term)
+    sin_cache[:, 0::2] = np.sin(position * div_term)
+    sin_cache[:, 1::2] = np.sin(position * div_term)
+
+    cos_initializer = numpy_helper.from_array(cos_cache, name="cos_cache")
+    sin_initializer = numpy_helper.from_array(sin_cache, name="sin_cache")
+
+    input_tensor = helper.make_tensor_value_info("input", TensorProto.FLOAT, [batch_size, seq_len, hidden_size])
+    position_ids_tensor = helper.make_tensor_value_info("position_ids", TensorProto.INT64, [batch_size, seq_len])
+    output_tensor = helper.make_tensor_value_info("output", TensorProto.FLOAT, [batch_size, seq_len, hidden_size])
+
+    nodes = [
+        helper.make_node(
+            OpType.RotaryEmbedding,
+            inputs=["input", "position_ids", "cos_cache", "sin_cache"],
+            outputs=["output"],
+            name="rotary_embedding_node",
+            domain=MSFT_DOMAIN,
+            interleaved=0,
+            num_heads=0,
+            rotary_embedding_dim=hidden_size,
+            scale=1.0,
+        ),
+    ]
+
+    graph = helper.make_graph(
+        nodes=nodes,
+        name="TestGraph",
+        inputs=[input_tensor, position_ids_tensor],
+        outputs=[output_tensor],
+        initializer=[cos_initializer, sin_initializer],
+    )
+    model = helper.make_model(graph, opset_imports=[helper.make_opsetid("", 14), helper.make_opsetid(MSFT_DOMAIN, 1)])
+    model.ir_version = 8
```

Comment:
Is opset 14 preferred, or would it be good to stick with 21 given that is what model builder uses?